### ISSUE TITLE
RC4. Animation. void => * animation is not fired on component detach

### ISSUE THREAD

**Issue by kamarouski** [angular/angular#10520]
**I'm submitting a ...**  (check one with "x")

```
[ x] bug report
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
It looks like \* => void animation is not fired when I try to detach a component through ViewContainerRef.detach().

I have the following animation defined for my component:

```
        trigger("appear", [
            transition('* => void', [
                style({ opacity: 1 }),
                animate(".2s cubic-bezier(0.25, 0.8, 0.25, 1)", style({opacity: 0}))
            ]),
            transition('void => *', [
                style({opacity: 0}),
                animate(".2s cubic-bezier(0.25, 0.8, 0.25, 1)", style({opacity: 1}))
            ])
        ])
```

I dynamically create the component using ViewContainerRef.createComponent() and entering animation works fine.

Then when I remove component using ViewContainerRef.detach() or ViewContainerRef.clear() leave animation is not fired.

**Expected/desired behavior**
Leave animation is fired on ViewContainerRef.detach()

Environment:
- Chrome: 52.0.2743.116
- OS: Windows 10 + Anniversary Update
- Angular: 2.0.0-rc4

**Comment by matsko** [comment]
This will be looked at after RC6 is out.

**Comment by bobrosoft** [comment]
@kamarouski, hey, did you find a solution? Need to create and remove arbitrary components dynamically and want to animate "enter" and "leave"...

**Comment by bobrosoft** [comment]
Oh, it's actually working as expected.

**Comment by piotr-gawlowski** [comment]
@bobrosoft How is this working?

this triggers my **:enter**
this.contentCmpRef = this._viewContainerRef.createComponent(componentFactory);


and later neither
this.contentCmpRef.destroy();
or 
this._viewContainerRef.remove() or this._viewContainerRef.detach()
triggers animation for **:leave**
And yes - I have _display: block_. But I don't know how this make any difference :)

**Comment by bobrosoft** [comment]
@piotr-gawlowski this._viewContainerRef.detach() should trigger it. If not, something wrong with your animation definitions (it's easy to have a mistake there, especially in animation phases). Re-check again and debug in devtool.

**Comment by piotr-gawlowski** [comment]
@bobrosoft hey, I include my plnk to verify https://plnkr.co/edit/xGzFIY01C18YbyVHdeK4?p=preview
Do you know why their is no animation on leave?

**Comment by bobrosoft** [comment]
@piotr-gawlowski I don't remember any ":enter" and ":leave" special words in official documentation.. Why don't you use "void => *" and "* => void" state-strings as documentation says?

**Comment by matsko** [comment]
@piotr-gawlowski `:enter` and `:leave` are aliases for `void => *` and `* => void` as of of 2.1.0.

**Comment by piotr-gawlowski** [comment]
@matsko Thanks for answer, I know that. 
And @bobrosoft It's right here https://angular.io/docs/ts/latest/guide/animations.html#!#example-entering-and-leaving and I'a using it exactly right that here
https://plnkr.co/edit/2HbP5N?p=preview :)

**Comment by bobrosoft** [comment]
@piotr-gawlowski sorry, it's Friday and I'm already with a bottle of good wine here :) Check that plunker I made for testing of enter/leave animation for some recent project http://plnkr.co/edit/MKhFhWtHxuUmI6e4uxm6?p=preview  And find the difference :) Don't forget to say what the issue was.

**Comment by bobrosoft** [comment]
Oh, just noticed -- it based on routes, not dynamic components creation. But in current project I have dynamic components creation and all works perfectly. Anyway, check `host` and `animations` sections, use same values in your code. I think you have issue somewhere there.

**Comment by piotr-gawlowski** [comment]
@bobrosoft Thanks for that example :)
It's perfect example how to use animation engine with Router. So - animate on route change. 
Which doesn't relates much to this issue which is focusing on programmatically created components.
Just wanted to know is it even possible to trigger leave animation on component "destroy". Or meybe I get whole thing wrong, But I guess It's @matsko how can answer this.

**Comment by bobrosoft** [comment]
Ok, I fixed your stuff https://plnkr.co/edit/nbLabGk2PoKiBoFNjGI4?p=preview %) Noticed `host` and where `@enter` moved? Right as in my example above. Meh. Can I go and drink today...

**Comment by robwormald** [comment]
what's happening here (I think, @matsko can correct me) is that the component you're detaching is being detached before the nested animations have a chance to run.

This seems like a reasonable use case for using a host animation binding, as presumably you actually want to animate the container leaving. See https://plnkr.co/edit/XCNePBaccTN9q9wJSWxo?p=preview

**Comment by bobrosoft** [comment]
@kamarouski hey, can you close that issue? It starts to be an stackoverflow

**Comment by bobrosoft** [comment]
@robwormald I fixed his example.

**Comment by bobrosoft** [comment]
@robwormald btw, do you paid to write and support ng2? Just wondering :) Can I join the team? :D

**Comment by piotr-gawlowski** [comment]
@robwormald @bobrosoft 
Thanks - this is exactly how this should work. Appreciated.

**Comment by lazarljubenovic** [comment]
This can be closed since https://github.com/angular/angular/issues/10520#issuecomment-263023880 is correct. The leaving animation will run when it's on the host, otherwise not (which is expected).

**Comment by matsko** [comment]
The main issue is that if the container removes itself then any sub animations will not get run because the parent always wins. Even if the parent doesn't have an animation then the children will not get a chance to animate.

But with yesterday's animation patch the callbacks should now fire on child elements even when the parent has no animation inflight during removal.

Also, now that `query` has landed, a parent animation can query for inner animations and let them run. Let's try putting together a demo in the next day or two once the release is out.

**Comment by matsko** [comment]
Closing this for now. Use `query(':leave', animateChild())` to animate inner child animations.
